modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/SleepTimer.kt

```diff
@@ -126,13 +129,28 @@ class SleepTimer @Inject constructor(
 
     private fun shouldRestartSleepEndOfChapter(diffTime: Duration, isSleepEndOfChapterRunning: Boolean) = diffTime < MIN_TIME_TO_RESTART_SLEEP_TIMER_IN_MINUTES && !isSleepEndOfChapterRunning && lastSleepAfterEndOfChapterTime != null
 
+    @RequiresApi(Build.VERSION_CODES.S)
+    private fun requestExactAlarmPermissionIfNeeded(context: Context) {
+        val intent = Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM)
+        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
+        context.startActivity(intent)
+    }
+
     private fun createAlarm(timeMs: Long): Boolean {
         val sleepIntent = getSleepIntent()
         val alarmManager = getAlarmManager()
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && !alarmManager.canScheduleExactAlarms()) {
+            LogBuffer.e(TAG, "Cannot schedule exact alarms. Permission not granted.")
+            requestExactAlarmPermissionIfNeeded(context)
+            return false
+        }
+
         alarmManager.cancel(sleepIntent)
+
         return try {
             LogBuffer.i(TAG, "Starting...")
-            alarmManager.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, timeMs, sleepIntent)
+            alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, timeMs, sleepIntent)
```

Comment:
I have been wondering if there is a way to do this without using the alarm permission. When playback happens, we have the playback service running in the foreground. This service saves the playback progress to the database. It could also check if the sleep timer should have been triggered and pause the podcast. I could be wrong about this. Is there a reason it could work like this?

Reply:
hummm... I don't know why we implemented this using system alarm, but I think we can try this approach. I am going to do some test so this how this could work